File: ali-backend/app/types/tutorial_lifecycle.py
"""
ALI Enterprise Application - Tutorial Lifecycle Types
Implements DRAFT → IN_REVIEW → PUBLISHED → ARCHIVED workflow per Spec v2.2 §7
"""
from enum import Enum
from typing import TypedDict, Optional, List, Any
from datetime import datetime


class TutorialStatus(str, Enum):
    """Lifecycle states for tutorial content."""
    DRAFT = "DRAFT"           # Generated + QC complete, not user-facing
    IN_REVIEW = "IN_REVIEW"   # Admin reviewing; may request regeneration
    PUBLISHED = "PUBLISHED"   # Available to learners; immutable version
    ARCHIVED = "ARCHIVED"     # Retained for audit/rollback; not active


class TutorialRequestStatus(str, Enum):
    """States for user tutorial requests."""
    PENDING = "PENDING"         # User submitted, awaiting admin
    APPROVED = "APPROVED"       # Admin approved, ready for generation
    DENIED = "DENIED"           # Admin denied the request
    GENERATING = "GENERATING"   # Generation in progress
    COMPLETED = "COMPLETED"     # Generation complete, tutorial created
    FAILED = "FAILED"           # Generation failed


class TutorialVersion(TypedDict):
    """Immutable version snapshot for rollback support."""
    versionId: str
    hash: str
    timestamp: str  # ISO format
    modelVersion: str
    publishedBy: Optional[str]


class RubricScore(TypedDict):
    """Individual rubric dimension score."""
    dimension: str
    score: int  # 0-100
    passed: bool
    issues: List[str]


class RubricReport(TypedDict):
    """Critic Agent rubric validation report."""
    verdict: str  # PASS or FAIL
    overallScore: int  # 0-100
    scores: List[RubricScore]
    citationCoverage: float  # 0.0-1.0
    cardWordCountViolations: int
    syntaxErrors: List[str]
    fixList: List[str]
    validatedAt: str


class EvidenceBundle(TypedDict):
    """Citation and source evidence for grounded content."""
    sources: List[dict]
    citations: List[dict]
    truthSourceJson: Optional[str]  # GCS path
    createdAt: str


class TutorialMetadata(TypedDict):
    """Full tutorial metadata with lifecycle support."""
    id: str
    title: str
    topic: str
    userId: str
    status: str  # TutorialStatus value
    versions: List[TutorialVersion]
    currentVersion: str
    rubricReport: Optional[RubricReport]
    evidenceBundle: Optional[EvidenceBundle]
    createdAt: str
    updatedAt: str
    publishedAt: Optional[str]


class TutorialRequest(TypedDict):
    """User's request for tutorial generation."""
    requestId: str
    userId: str
    userEmail: str
    topic: str
    context: Optional[str]
    status: str  # TutorialRequestStatus value
    createdAt: Any  # Firestore timestamp
    adminDecision: Optional[dict]
    tutorialId: Optional[str]  # Set after generation completes

File: ali-frontend/src/types/tutorial-lifecycle.ts
/**
 * ALI Enterprise Application - Tutorial Lifecycle Types
 * Spec v2.2 Tutorial Generation §7: Draft → Review → Publish Workflow
 */

export type TutorialStatus = 'DRAFT' | 'IN_REVIEW' | 'PUBLISHED' | 'ARCHIVED';

export type TutorialRequestStatus =
    | 'PENDING'
    | 'APPROVED'
    | 'DENIED'
    | 'GENERATING'
    | 'COMPLETED'
    | 'FAILED';

export interface TutorialVersion {
    versionId: string;
    hash: string;
    timestamp: string;
    modelVersion: string;
    publishedBy?: string;
}

export interface RubricScore {
    dimension: string;
    score: number;
    passed: boolean;
    issues: string[];
}

export interface RubricReport {
    verdict: 'PASS' | 'FAIL';
    overallScore: number;
    scores: RubricScore[];
    citationCoverage: number;
    cardWordCountViolations: number;
    syntaxErrors: string[];
    fixList: string[];
    validatedAt: string;
}

export interface TutorialRequest {
    requestId: string;
    userId: string;
    userEmail: string;
    topic: string;
    context?: string;
    status: TutorialRequestStatus;
    createdAt: Date;
    adminDecision?: {
        action: 'approved' | 'denied';
        approvedBy?: string;
        deniedBy?: string;
        approvedAt?: Date;
        deniedAt?: Date;
        reason?: string;
    };
    tutorialId?: string;
}
